### PR TITLE
Integrate with JWT

### PR DESCRIPTION
**Pull request by dmitry-silchenko**
Sample of the code:

```
const mailingClient = new MailingClient('https://api-integration.bundesimmo.de/mailing-service', {
  serviceToken: {
    // jwt: {
    //   accessId: 'mailingservice',
    //   secretKey: 'ZT...=='
    // },
    // Either
    authorization: () => {
      const accessId = 'mailingservice'
      const secretKey = 'ZT...=='

      const jwt = require('jsonwebtoken');
      const signature = jwt.sign({}, secretKey, { algorithm: 'HS256', expiresIn: '1m', issuer: accessId });

      return `JWT ${accessId}:${signature}`
    }
  }
})
```

We can either provide access_id and secret_key to generate JWT or we can provide `authorization` function (might be async) which generates and returns a string for `authorization` header.

### CODE REVIEW

File: src/clients/base-client.js [hunk 1]
```diff
@@ -163,6 +184,7 @@ class BaseClient {
     const url = buildUrl(this.baseUrl, path, parameters)
 
     return this.sendRequest(url, {
+      body: parameters.data,
```
Comment (dmitry-silchenko) [inline src/clients/base-client.js]: I wish `delete` initially accepted `data` as a parameter, but... I guess introducing it now can break some apps (new parameter which is not the last one)

File: package.json [hunk 2]
```diff
@@ -21,7 +21,8 @@
     "camelcase-keys": "^6.1.1",
     "mime": "^2.3.1",
     "qs": "^6.8.3",
-    "uuid": "^3.2.1"
+    "uuid": "^3.2.1",
+    "jsonwebtoken": "^9.0.2"
```
Comment (jdahlke) [inline package.json]: I would not add `jsonwebtoken` to dependencies. It's not used in every context and it's not small.

We could add text to the README, that `jsonwebtoken` needs to be installed in the host application.
The README needs some updates anyway.

File: src/clients/base-client.js [hunk 3]
```diff
@@ -38,7 +38,28 @@ class BaseClient {
       { baseUrl: Config.serviceTokenUrl },
       options.serviceToken
     )
-    this.tokenClient = new SharkProxy.ServiceTokenClient(serviceTokenOptions)
+
+    if (serviceTokenOptions.authorization) {
```
Comment (jdahlke) [inline src/clients/base-client.js]: I would rename `serviceTokenOptions.authorization` to `authorizationFunc` or `authFunc`.
`serviceTokenOptions.authorization` does not make me think of a function. Or you need to make a typecheck for this parameter.

Comment (jdahlke) [reply src/clients/base-client.js]: And to finish what you already started, I would move like to move the authorization function outside of this client. Only keep something for backward compatibility.

my idea
```
if (options.authorizationFunc) {
  // the "new" way with function composition
  this.authorizationFunc = async () => await options.authorizationFunc()
  // is this the same as 
  // this.authorizationFunc = options.authorizationFunc ?
} else {
  // for downward com
  const tokenClient = new SharkProxy.ServiceTokenClient({
    Object.assign({ baseUrl: Config.serviceTokenUrl },  options.serviceToken)
  })

  this.authorizationFunc = async () => {
    const token = await tokenClient.createServiceToken({})

    return `Bearer ${token.jwt}`
  }
}
```

``` 
// new file and function is exported in `src/shark.js`

function jwtAuthorization({ accessId, secretKey }) {
  const jwt = require('jsonwebtoken')
  const signature = jwt.sign({}, secretKey, { algorithm: 'HS256', expiresIn: '1m', issuer: accessId })

  return `JWT ${accessId}:${signature}`
}
```

Comment (jdahlke) [reply src/clients/base-client.js]: The usage would then be like (copied from your pull request description)

```
const { MailingClient, jwtAuthorization } from require('bima-shark-sdk')

const authorizationFunc = async () => {
   const accessId = 'mailingservice'
   const secretKey = 'ZT...=='

   return jwtAuthorization({ accessId, secretKey })
}

const mailingClient = new MailingClient('https://api-integration.bundesimmo.de/mailing-service', {
    authorizationFunc: authorizationFunc
  }
})
...
```

File: package.json (hunk 2 above)
Comment (dmitry-silchenko) [reply package.json]: so that `jwt` option in params might be overkill. let's leave `authFunc` only.

File: src/clients/base-client.js (hunk 3 above)
Comment (dmitry-silchenko) [reply src/clients/base-client.js]: > this.authorizationFunc = async () => await options.authorizationFunc()
>   // is this the same as 
>   // this.authorizationFunc = options.authorizationFunc ?

Almost, in first case it is more 'isolated' which means `this` cannot be used inside user's function. Additionally, there is one trick - we make `authorizationFunc` async and do not care if original, specified by user is async as well or not, we can expect a promise and append request with `.then` - again regardless if user's function is async or not.

Comment (dmitry-silchenko) [reply src/clients/base-client.js]: > I would rename `serviceTokenOptions.authorization` to `authorizationFunc` or `authFunc`.

what about `getToken` ?

Comment (jdahlke) [reply src/clients/base-client.js]: Good idea, but I would use`getAuthToken`. Because a token could mean a lot depending on context.

**Review by jdahlke — APPROVED** [review]
Maybe rename `getToken` to `getAuthToken`, but looks good so far.

**Review by jdahlke — APPROVED** [review]

**Pull request merged**